// ciscospark/dist/plugins/team-memberships.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _page = require('../lib/page');

var _page2 = _interopRequireDefault(_page);

var _sparkCore = require('@ciscospark/spark-core');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @typedef {Object} Types~TeamMembership
 * @property {string} id - (server generated) The team ID
 * @property {string} personId - The person ID
 * @property {string} personEmail - The email address of the person
 * @property {boolean} isModerator - Set to `true` to make the person a team
 * moderator
 */

/**
 * Team Memberships represent a person's relationship to a team. Use this API to
 * list members of any team that you're in or create memberships to invite
 * someone to a team. Team memberships can also be updated to make someome a
 * moderator or deleted to remove them from the team.
 *
 * Just like in the Spark app, you must be a member of the team in order to list
 * its memberships or invite people.
 */
/**!
 *
 * Copyright (c) 2015-2016 Cisco Systems, Inc. See LICENSE file.
 * @private
 */

var TeamMemberships = _sparkCore.SparkPlugin.extend({
  /**
   * Add someone to a team by Person ID or email address; optionally making them
   * a moderator.
   * @param {Types~TeamMembership} membership
   * @returns {Promise<Types~TeamMembership>}
   * @memberof TeamMemberships.prototype
   * @example
   * <%= team_memberships__create %>
   */

  create: function create(membership) {
    return this.request({
      method: 'POST',
      uri: this.config.hydraServiceUrl + '/team/memberships',
      body: membership
    }).then(function (res) {
      return res.body;
    });
  },


  /**
   * Get details for a membership by ID.
   * @param {Types~TeamMembership|string} membership
   * @returns {Promise<Types~TeamMembership>}
   * @example
   * <%= team_memberships__get %>
   */
  get: function get(membership) {
    var id = membership.id || membership;
    return this.request({
      uri: this.config.hydraServiceUrl + '/team/memberships/' + id
    }).then(function (res) {
      return res.body.items || res.body;
    });
  },


  /**
   * Lists all team memberships. By default, lists memberships for teams to
   * which the authenticated user belongs.
   * @param {Object} options
   * @param {string} options.max
   * @returns {[type]}
   * @example
   * <%= team_memberships__list %>
   */
  list: function list(options) {
    var _this = this;

    return this.request({
      uri: this.config.hydraServiceUrl + '/team/memberships',
      qs: options
    }).then(function (res) {
      return new _page2.default(res, _this.spark);
    });
  },


  /**
   * Deletes a membership by ID.
   * @param {Types~TeamMembership|string} membership
   * @returns {Promise}
   * @example
   * <%= team_memberships__remove %>
   */
  remove: function remove(membership) {
    var id = membership.id || membership;

    return this.request({
      method: 'DELETE',
      uri: this.config.hydraServiceUrl + '/team/memberships/' + id
    }).then(function (res) {
      // Firefox has some issues with 204s and/or DELETE. This should move to
      // http-core
      if (res.statusCode === 204) {
        return undefined;
      }
      return res.body;
    });
  },


  /**
   * Updates properties for a membership.
   * @param {Types~TeamMembership} membership
   * @returns {Promise<Types~TeamMembership>}
   * @example
   * <%= team_memberships__update %>
   */
  update: function update(membership) {
    var id = membership.id || membership;
    return this.request({
      method: 'PUT',
      uri: this.config.hydraServiceUrl + '/team/memberships/' + id,
      body: membership
    }).then(function (res) {
      return res.body;
    });
  }
});

exports.default = TeamMemberships;
